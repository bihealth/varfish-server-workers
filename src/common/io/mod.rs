//! Common, IO-related code.

pub mod std;
