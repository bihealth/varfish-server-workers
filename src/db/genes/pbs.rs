//! Data structures for (de-)serialization as generated by `prost-build`.

include!(concat!(env!("OUT_DIR"), "/varfish.v1.gene.rs"));
