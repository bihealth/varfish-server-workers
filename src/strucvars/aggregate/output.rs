//! Code for writing the output file.

use mehari::annotate::strucvars::{
    bnd::Breakend, csq::interface::StrandOrientation, PeOrientation,
};
use noodles_vcf as vcf;

use crate::{
    common::{Chrom, Genotype},
    strucvars::query::schema::SvType,
};

/// Representation of the fields for the in-house background database.
#[derive(Debug, Clone, serde::Deserialize, serde::Serialize)]
pub struct Record {
    /// chromosome name
    pub chromosome: String,
    /// start position, 1-based
    pub begin: i32,
    /// chromosome2 name
    pub chromosome2: String,
    /// end position, 1-based
    pub end: i32,
    /// paired-end orientation
    pub pe_orientation: StrandOrientation,
    /// type of the SV
    pub sv_type: SvType,
    /// number of overall carriers
    pub carriers: u32,
    /// number of het. carriers
    pub carriers_het: u32,
    /// number of hom. carriers
    pub carriers_hom: u32,
    /// number of hemi. carriers
    pub carriers_hemi: u32,
}

impl Record {
    /// Compute reciprocal overlap between `self` and `other`.
    pub fn overlap(&self, other: &Record) -> f32 {
        let s1 = if self.begin > 0 { self.begin - 1 } else { 0 };
        let e1 = self.end + 1;
        let s2 = if other.begin > 0 { other.begin - 1 } else { 0 };
        let e2 = other.end;

        let ovl_s = std::cmp::max(s1, s2);
        let ovl_e = std::cmp::min(e1, e2);
        if ovl_e <= ovl_s {
            0.0
        } else {
            let len1 = (e1 - s1) as f32;
            let len2 = (e2 - s2) as f32;
            let ovl_len = (ovl_e - ovl_s) as f32;
            (ovl_len / len1).min(ovl_len / len2)
        }
    }

    pub fn merge_into(&mut self, other: &Record) {
        self.carriers += other.carriers;
        self.carriers_het += other.carriers_het;
        self.carriers_hom += other.carriers_hom;
        self.carriers_hemi += other.carriers_hemi;
    }

    /// Convert VCF record into a `Record`.
    pub fn from_vcf(
        record: &vcf::Record,
        header: &vcf::Header,
        _genome_release: crate::common::GenomeRelease,
        pedigree: &mehari::ped::PedigreeByName,
    ) -> Result<Self, anyhow::Error> {
        let chromosome = record.chromosome().to_string();
        let begin = {
            let position: usize = record.position().into();
            position.saturating_sub(1) as i32
        };
        let chromosome2 = if let Some(Some(vcf::record::info::field::Value::String(chromosome2))) =
            record.info().get(
                &"chr2"
                    .parse::<vcf::record::info::field::Key>()
                    .expect("invalid key chr2?"),
            ) {
            chromosome2.clone()
        } else {
            chromosome.clone()
        };
        let end = if let Some(Some(vcf::record::info::field::Value::Integer(end))) = record
            .info()
            .get(&vcf::record::info::field::key::END_POSITION)
        {
            *end
        } else {
            anyhow::bail!("missing INFO/END")
        };

        let sv_type = if let Some(Some(vcf::record::info::field::Value::String(sv_type))) =
            record.info().get(&vcf::record::info::field::key::SV_TYPE)
        {
            sv_type.parse()?
        } else {
            anyhow::bail!("missing INFO/SVTYPE")
        };

        let pe_orientation = match sv_type {
            SvType::Del => StrandOrientation::ThreeToFive,
            SvType::Dup => StrandOrientation::FiveToThree,
            SvType::Inv => StrandOrientation::FiveToFive,
            SvType::Bnd => {
                let pe_orientation = Breakend::from_ref_alt_str(
                    &record.reference_bases().to_string(),
                    &record
                        .alternate_bases()
                        .first()
                        .ok_or_else(|| anyhow::anyhow!("no alternate allele?"))?
                        .to_string(),
                )?
                .pe_orientation;
                match pe_orientation {
                    PeOrientation::ThreeToThree => StrandOrientation::ThreeToThree,
                    PeOrientation::FiveToFive => StrandOrientation::FiveToFive,
                    PeOrientation::ThreeToFive => StrandOrientation::ThreeToFive,
                    PeOrientation::FiveToThree => StrandOrientation::FiveToThree,
                    PeOrientation::Other => StrandOrientation::NotApplicable,
                }
            }
            SvType::Ins | SvType::Cnv => StrandOrientation::NotApplicable,
        };

        let chrom: Chrom =
            annonars::common::cli::canonicalize(record.chromosome().to_string().as_str())
                .as_str()
                .parse()?;

        let (mut carriers_het, mut carriers_hom, mut carriers_hemi) = (0, 0, 0);
        for (name, sample) in header
            .sample_names()
            .iter()
            .zip(record.genotypes().values())
        {
            let individual = pedigree
                .individuals
                .get(name)
                .ok_or_else(|| anyhow::anyhow!("individual {} not found in pedigree", name))?;

            let genotype: Genotype =
                if let Some(Some(gt)) = sample.get(&vcf::record::genotypes::keys::key::GENOTYPE) {
                    if let vcf::record::genotypes::sample::Value::String(gt) = gt {
                        gt.as_str().parse()?
                    } else {
                        anyhow::bail!("invalid genotype value: {:?}", gt);
                    }
                } else {
                    continue; // skip, no-call or empty
                };

            match (chrom, individual.sex, genotype) {
                (_, _, Genotype::WithNoCall) => continue,
                // on the autosomes, male/female count the same
                (Chrom::Auto, _, Genotype::HomRef) => (),
                (Chrom::Auto, _, Genotype::Het) => {
                    carriers_het += 1;
                }
                (Chrom::Auto, _, Genotype::HomAlt) => {
                    carriers_hom += 1;
                }
                // on the gonomosomes, we handle call male variant calls as hemizygous
                (Chrom::X, mehari::ped::Sex::Male, Genotype::HomRef)
                | (Chrom::Y, mehari::ped::Sex::Male, Genotype::HomRef) => (),
                (Chrom::X, mehari::ped::Sex::Male, Genotype::HomAlt)
                | (Chrom::X, mehari::ped::Sex::Male, Genotype::Het)
                | (Chrom::Y, mehari::ped::Sex::Male, Genotype::Het)
                | (Chrom::Y, mehari::ped::Sex::Male, Genotype::HomAlt) => {
                    carriers_hemi += 1;
                }
                // for female samples, we handle chrX as biallelic
                (Chrom::X, mehari::ped::Sex::Female, Genotype::HomRef)
                | (Chrom::X, mehari::ped::Sex::Female, Genotype::Het) => {
                    carriers_het += 1;
                }
                (Chrom::X, mehari::ped::Sex::Female, Genotype::HomAlt) => {
                    carriers_hom += 1;
                }
                // we ignore calls to chrY for female samples
                (Chrom::Y, mehari::ped::Sex::Female, Genotype::HomRef)
                | (Chrom::Y, mehari::ped::Sex::Female, Genotype::Het)
                | (Chrom::Y, mehari::ped::Sex::Female, Genotype::HomAlt) => (),
                // do not count samples with unknown sex on gonomosomes
                (Chrom::X, mehari::ped::Sex::Unknown, _)
                | (Chrom::Y, mehari::ped::Sex::Unknown, _) => (),
            };
        }

        Ok(Self {
            chromosome,
            begin,
            chromosome2,
            end,
            pe_orientation,
            sv_type,
            carriers_het,
            carriers_hom,
            carriers_hemi,
            carriers: carriers_het + carriers_hom + carriers_hemi,
        })
    }
}
