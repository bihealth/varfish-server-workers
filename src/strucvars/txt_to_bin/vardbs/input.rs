//! Code supporting the I/O of public database records and a common
//! `InputRecord` for common representation.

use serde::Deserialize;
use tracing::error;

use crate::strucvars::aggregate::output::Record as InhouseDbRecord;
use crate::strucvars::query::schema::SvType;

/// dbVar database record as read from TSV file.
#[derive(Debug, Deserialize)]
pub struct DbVarRecord {
    /// chromosome name
    pub chromosome: String,
    /// begin position, 0-based
    pub begin: i32,
    /// end position, 0-based
    pub end: i32,
    /// number of overall carriers
    pub num_carriers: u32,
    /// type of the SV
    pub sv_type: String,
}

/// DGV database record as read from TSV file.
#[derive(Debug, Deserialize)]
pub struct DgvRecord {
    /// chromosome name
    pub chromosome: String,
    /// begin position, 0-based
    pub begin: i32,
    /// end position, 0-based
    pub end: i32,
    /// The structural variant type
    sv_type: String,
    /// Number of observed gains.
    observed_gains: u32,
    /// Number of observed losses
    observed_losses: u32,
}

/// DGV gold standard database record as read from TSV file.
#[derive(Debug, Deserialize)]
pub struct DgvGsRecord {
    /// chromosome name
    pub chromosome: String,
    /// begin position, 0-based
    pub begin_outer: i32,
    /// outer end position, 0-based
    pub end_outer: i32,
    /// The structural variant type
    pub sv_sub_type: String,
    /// Number of carriers.
    pub num_carriers: u32,
}

/// ExAC CNV database record as read from TSV file for deserialization
/// from TSV.
#[derive(Deserialize, Debug)]
pub struct ExacRecord {
    /// chromosome name
    pub chromosome: String,
    /// outer start position, 0-based
    pub begin: i32,
    /// outer end position, 0-based
    pub end: i32,
    /// The structural vairant type
    pub sv_type: String,
}

/// Thousand Genomes SV database record as read from TSV file.
#[derive(Debug, Deserialize)]
pub struct G1kRecord {
    /// chromosome name
    pub chromosome: String,
    /// begin position, 0-based
    pub begin: i32,
    /// end position, 0-based
    pub end: i32,
    /// The structural vairant type
    pub sv_type: String,
    /// Number of hom. alt. alleles.
    pub n_homalt: u32,
    /// Number of het. alt. alleles.
    pub n_het: u32,
}

/// gnomAD SV v2 database record as read from TSV file.
#[derive(Debug, Deserialize)]
pub struct GnomadSv2Record {
    /// chromosome name
    pub chromosome: String,
    /// begin position, 0-based
    pub begin: i32,
    /// end position, 0-based
    pub end: i32,
    /// The structural vairant type
    pub svtype: String,
    /// Number of homozygous alternative carriers
    pub n_homalt: u32,
    /// Number of heterozygous carriers
    pub n_het: u32,
}

/// gnomAD SV v4 database record as read from TSV file.
#[derive(Debug, Deserialize)]
pub struct GnomadSv4Record {
    /// chromosome name
    pub chromosome: String,
    /// begin position, 0-based
    pub begin: i32,
    /// end position, 0-based
    pub end: i32,
    /// The structural vairant type
    pub svtype: String,
    /// Number of male homozygous reference allele carriers.
    pub male_n_homref: u32,
    /// Number of male heterozygous alternate allele carriers.
    pub male_n_het: u32,
    /// Number of male homozygous alternate allele carriers.
    pub male_n_homalt: u32,
    /// Number of male hemizygous alternate allele carriers.
    pub male_n_hemiref: u32,
    /// Number of male hemizygous reference allele carriers.
    pub male_n_hemialt: u32,
    /// Number of female homozygous reference allele carriers.
    pub female_n_homref: u32,
    /// Number of female heterozygous alternate allele carriers.
    pub female_n_het: u32,
    /// Number of female homozygous alternate allele carriers.
    pub female_n_homalt: u32,
    /// Number of samples at this site (CNV only).
    pub cnv_n_total: u32,
    /// Number of samples with a CNV at this site (CNV only).
    pub cnv_n_var: u32,
}

/// gnomAD CNV v$ database record as read from TSV file.
#[derive(Debug, Deserialize)]
pub struct GnomadCnv4Record {
    /// chromosome name
    pub chromosome: String,
    /// begin position, 0-based
    pub begin: i32,
    /// end position, 0-based
    pub end: i32,
    /// The structural vairant type
    pub svtype: String,
    /// Number of samples at this site (passing QC).
    pub n_total: u32,
    /// Number of samples with a CNV at this site (passing QC).
    pub n_var: u32,
}

/// Common type to convert input data to.
pub struct InputRecord {
    /// Chromosome of start position.
    pub chromosome: String,
    /// Chromosome of end position.
    pub chromosome2: String,
    /// SV type
    pub sv_type: SvType,
    /// 0-based begin position
    pub begin: i32,
    /// 0-based end position
    pub end: i32,
    /// Number of carriers (or alleles), depending on database.
    pub count: u32,
}

impl TryInto<Option<InputRecord>> for InhouseDbRecord {
    type Error = &'static str;

    fn try_into(self) -> Result<Option<InputRecord>, Self::Error> {
        Ok(Some(InputRecord {
            chromosome: self.chromosome,
            chromosome2: self.chromosome2,
            sv_type: self.sv_type,
            begin: self.begin,
            end: self.end,
            count: self.carriers,
        }))
    }
}

impl TryInto<Option<InputRecord>> for DbVarRecord {
    type Error = &'static str;

    fn try_into(self) -> Result<Option<InputRecord>, Self::Error> {
        let sv_type = match self.sv_type.split(';').next().unwrap() {
            "alu_insertion"
            | "herv_insertion"
            | "insertion"
            | "line1_insertion"
            | "mobile_element_insertion"
            | "novel_sequence_insertion"
            | "sva_insertion" => SvType::Ins,
            "copy_number_gain" | "duplication" | "tandem_duplication" => SvType::Dup,
            "alu_deletion" | "copy_number_loss" | "deletion" | "herv_deletion"
            | "line1_deletion" | "sva_deletion" => SvType::Del,
            "copy_number_variation" => SvType::Cnv,
            _ => {
                error!("sv_type = {}", &self.sv_type);
                return Err("unknown SV type");
            }
        };
        Ok(Some(InputRecord {
            chromosome: self.chromosome.clone(),
            chromosome2: self.chromosome,
            begin: self.begin,
            end: self.end,
            sv_type,
            count: 1,
        }))
    }
}

impl TryInto<Option<InputRecord>> for DgvRecord {
    type Error = &'static str;

    fn try_into(self) -> Result<Option<InputRecord>, Self::Error> {
        let sv_type = match self.sv_type.as_ref() {
            "alu deletion"
            | "deletion"
            | "herv deletion"
            | "line1 deletion"
            | "mobile element deletion"
            | "loss"
            | "sva deletion" => SvType::Del,
            "alu insertion"
            | "herv insertion"
            | "insertion"
            | "line1 insertion"
            | "mobile element insertion"
            | "novel sequence insertion"
            | "sva insertion" => SvType::Ins,
            "duplication" | "gain" | "tandem duplication" => SvType::Dup,
            "sequence alteration" | "complex" => return Ok(None), // skip
            "gain+loss" | "CNV" => SvType::Cnv,
            "inversion" => SvType::Inv,
            "OTHER" => return Ok(None), // skip
            _ => {
                error!("sv_type = {}", &self.sv_type);
                return Err("unknown SV type");
            }
        };
        Ok(Some(InputRecord {
            chromosome: self.chromosome.clone(),
            chromosome2: self.chromosome,
            begin: self.begin,
            end: self.end,
            sv_type,
            count: self.observed_gains + self.observed_losses,
        }))
    }
}

impl TryInto<Option<InputRecord>> for DgvGsRecord {
    type Error = &'static str;

    fn try_into(self) -> Result<Option<InputRecord>, Self::Error> {
        let sv_type = match self.sv_sub_type.as_ref() {
            "Gain" => SvType::Dup,
            "Loss" => SvType::Del,
            _ => {
                error!("sv_type = {}", &self.sv_sub_type);
                return Err("unknown SV type");
            }
        };
        Ok(Some(InputRecord {
            chromosome: self.chromosome.clone(),
            chromosome2: self.chromosome,
            begin: self.begin_outer,
            end: self.end_outer,
            sv_type,
            count: self.num_carriers,
        }))
    }
}

impl TryInto<Option<InputRecord>> for ExacRecord {
    type Error = &'static str;

    fn try_into(self) -> Result<Option<InputRecord>, Self::Error> {
        let sv_type = match self.sv_type.as_ref() {
            "DUP" => SvType::Dup,
            "DEL" => SvType::Del,
            _ => {
                error!("sv_type = {}", &self.sv_type);
                return Err("unknown SV type");
            }
        };
        Ok(Some(InputRecord {
            chromosome: self.chromosome.clone(),
            chromosome2: self.chromosome,
            begin: self.begin,
            end: self.end,
            sv_type,
            count: 1,
        }))
    }
}

impl TryInto<Option<InputRecord>> for GnomadSv2Record {
    type Error = &'static str;

    fn try_into(self) -> Result<Option<InputRecord>, Self::Error> {
        let sv_type = match self.svtype.as_str() {
            "CPX" => return Ok(None), // no correspondence
            "CTX" | "BND" => SvType::Bnd,
            "DEL" => SvType::Del,
            "DUP" => SvType::Dup,
            "INS" => SvType::Ins,
            "INV" => SvType::Inv,
            "MCNV" => SvType::Cnv,
            _ => {
                error!("sv_type = {}", &self.svtype);
                return Err("unknown SV type");
            }
        };
        Ok(Some(InputRecord {
            chromosome: self.chromosome.clone(),
            chromosome2: self.chromosome,
            begin: self.begin - 1,
            end: self.end,
            sv_type,
            count: self.n_homalt + self.n_het,
        }))
    }
}

impl TryInto<Option<InputRecord>> for GnomadCnv4Record {
    type Error = &'static str;

    fn try_into(self) -> Result<Option<InputRecord>, Self::Error> {
        Ok(Some(InputRecord {
            chromosome: self.chromosome.clone(),
            chromosome2: self.chromosome,
            begin: self.begin,
            end: self.end,
            sv_type: match self.svtype.as_str() {
                "DEL" => SvType::Del,
                "DUP" => SvType::Dup,
                _ => {
                    error!("sv_type = {}", &self.svtype);
                    return Err("unknown SV type");
                }
            },
            count: self.n_var,
        }))
    }
}

impl TryInto<Option<InputRecord>> for GnomadSv4Record {
    type Error = &'static str;

    fn try_into(self) -> Result<Option<InputRecord>, Self::Error> {
        Ok(Some(InputRecord {
            chromosome: self.chromosome.clone(),
            chromosome2: self.chromosome,
            begin: self.begin,
            end: self.end,
            sv_type: match self.svtype.as_str() {
                "BND" => SvType::Bnd,
                "CNV" => SvType::Cnv,
                "DEL" => SvType::Del,
                "DUP" => SvType::Dup,
                "INS" => SvType::Ins,
                "INV" => SvType::Inv,
                _ => {
                    error!("sv_type = {}", &self.svtype);
                    return Err("unknown SV type");
                }
            },
            count: self.male_n_het
                + self.male_n_homalt
                + self.male_n_hemialt
                + self.female_n_het
                + self.female_n_homalt
                + self.cnv_n_var,
        }))
    }
}

impl TryInto<Option<InputRecord>> for G1kRecord {
    type Error = &'static str;

    fn try_into(self) -> Result<Option<InputRecord>, Self::Error> {
        let sv_type = match self.sv_type.as_str() {
            "CN0" | "CNV" => SvType::Cnv,
            "DEL" => SvType::Del,
            "DEL_ALU" | "DEL_HERV" | "DEL_LINE1" | "DEL_SVA" => SvType::Del,
            "DUP" => SvType::Dup,
            "INV" => SvType::Inv,
            "INS" | "INS:ME:ALU" | "INS:ME:LINE1" | "INS:ME:SVA" => SvType::Ins,
            _ => {
                error!("sv_type = {}", &self.sv_type);
                return Err("unknown SV type");
            }
        };
        Ok(Some(InputRecord {
            chromosome: self.chromosome.clone(),
            chromosome2: self.chromosome,
            begin: self.begin,
            end: self.end,
            sv_type,
            count: self.n_homalt + self.n_het,
        }))
    }
}
