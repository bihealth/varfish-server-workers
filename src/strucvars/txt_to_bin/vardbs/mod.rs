//! Code for converting other structural variant database to binary (incl. in-house).

use std::fs::File;
use std::io::Write;
use std::path::Path;
use std::time::Instant;

use anyhow::anyhow;
use prost::Message;
use thousands::Separable;

use crate::common::{build_chrom_map, trace_rss_now};
use crate::pbs::svs::{BackgroundDatabase, BgDbRecord};
use crate::strucvars::aggregate::output::Record as InhouseDbRecord;
use crate::strucvars::query::schema::SvType;

use self::input::InputRecord;

mod input;

/// Known input file types.
#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Debug)]
pub enum InputFileType {
    Dbvar,
    Dgv,
    DgvGs,
    Exac,
    G1k,
    GnomadSv2,
    GnomadCnv4,
    GnomadSv4,
    InhouseDb,
}
/// Deserialize from CSV reader to an `Option<records::InputRecord>`
fn deserialize_loop<Rec>(
    reader: &mut csv::Reader<Box<dyn std::io::BufRead>>,
) -> Result<Vec<BgDbRecord>, anyhow::Error>
where
    Rec: core::fmt::Debug + TryInto<Option<InputRecord>> + for<'de> serde::Deserialize<'de>,
    <Rec as TryInto<std::option::Option<InputRecord>>>::Error: core::fmt::Debug,
    <Rec as TryInto<std::option::Option<InputRecord>>>::Error: Send,
    <Rec as TryInto<std::option::Option<InputRecord>>>::Error: std::marker::Sync,
{
    let chrom_map = build_chrom_map();
    let mut result = Vec::new();

    for record in reader.deserialize() {
        let record: Rec = record?;
        let maybe_record: Option<InputRecord> = record
            .try_into()
            .map_err(|err| anyhow!("problem with parsing: {:?}", &err))?;
        if let Some(record) = maybe_record {
            result.push(BgDbRecord {
                chrom_no: *chrom_map
                    .get(&record.chromosome)
                    .unwrap_or_else(|| panic!("unknown chrom: {:?}", &record.chromosome))
                    as i32,
                chrom_no2: *chrom_map
                    .get(&record.chromosome2)
                    .unwrap_or_else(|| panic!("unknown chrom2: {:?}", &record.chromosome2))
                    as i32,
                sv_type: match record.sv_type {
                    SvType::Del => crate::pbs::svs::SvType::Del,
                    SvType::Dup => crate::pbs::svs::SvType::Dup,
                    SvType::Inv => crate::pbs::svs::SvType::Inv,
                    SvType::Ins => crate::pbs::svs::SvType::Ins,
                    SvType::Bnd => crate::pbs::svs::SvType::Bnd,
                    SvType::Cnv => crate::pbs::svs::SvType::Cnv,
                } as i32,
                start: record.begin + 1,
                stop: record.end,
                count: record.count,
            });
        }
    }

    Ok(result)
}

/// Branch around `deserialize_loop`.
pub fn deserialize_branch(
    input_type: InputFileType,
    reader: &mut csv::Reader<Box<dyn std::io::BufRead>>,
) -> Result<Vec<BgDbRecord>, anyhow::Error> {
    match input_type {
        InputFileType::Dbvar => deserialize_loop::<input::DbVarRecord>(reader),
        InputFileType::Dgv => deserialize_loop::<input::DgvRecord>(reader),
        InputFileType::DgvGs => deserialize_loop::<input::DgvGsRecord>(reader),
        InputFileType::Exac => deserialize_loop::<input::ExacRecord>(reader),
        InputFileType::G1k => deserialize_loop::<input::G1kRecord>(reader),
        InputFileType::InhouseDb => deserialize_loop::<InhouseDbRecord>(reader),
        InputFileType::GnomadSv2 => deserialize_loop::<input::GnomadSv2Record>(reader),
        InputFileType::GnomadCnv4 => deserialize_loop::<input::GnomadCnv4Record>(reader),
        InputFileType::GnomadSv4 => deserialize_loop::<input::GnomadSv4Record>(reader),
    }
}

/// Perform conversion to protobuf `.bin` file.
pub fn convert_to_bin<P, Q>(
    path_input_tsv: P,
    path_output: Q,
    input_type: InputFileType,
) -> Result<(), anyhow::Error>
where
    P: AsRef<Path>,
    Q: AsRef<Path>,
{
    // Setup CSV reader for BED file - header is written as comment and must be
    // ignored.
    let mut reader = csv::ReaderBuilder::new()
        .has_headers(false)
        .comment(Some(b'#'))
        .delimiter(b'\t')
        .from_reader(mehari::common::io::std::open_read_maybe_gz(
            path_input_tsv.as_ref(),
        )?);
    let before_parsing = Instant::now();

    let records = deserialize_branch(input_type, &mut reader)?;
    let bg_db = BackgroundDatabase { records };

    tracing::debug!(
        "total time spent reading {} records: {:?}",
        bg_db.records.len().separate_with_commas(),
        before_parsing.elapsed()
    );
    trace_rss_now();

    let before_writing = Instant::now();
    let mut output_file = File::create(&path_output)?;
    output_file.write_all(&bg_db.encode_to_vec())?;
    output_file.sync_all()?;
    tracing::debug!(
        "total time spent writing {} records: {:?}",
        bg_db.records.len().separate_with_commas(),
        before_writing.elapsed()
    );

    Ok(())
}

#[cfg(test)]
mod test {
    use super::InputFileType;

    #[rstest::rstest]
    #[case::dbvar(
        InputFileType::Dbvar,
        "tests/db/to-bin/varfish-db-downloader/vardbs/grch37/strucvar/dbvar.bed.gz"
    )]
    #[case::dgv(
        InputFileType::Dgv,
        "tests/db/to-bin/varfish-db-downloader/vardbs/grch37/strucvar/dgv.bed.gz"
    )]
    #[case::dgv_gs(
        InputFileType::DgvGs,
        "tests/db/to-bin/varfish-db-downloader/vardbs/grch37/strucvar/dgv_gs.bed.gz"
    )]
    #[case::exac(
        InputFileType::Exac,
        "tests/db/to-bin/varfish-db-downloader/vardbs/grch37/strucvar/exac.bed.gz"
    )]
    #[case::g1k(
        InputFileType::G1k,
        "tests/db/to-bin/varfish-db-downloader/vardbs/grch37/strucvar/g1k.bed.gz"
    )]
    #[case::gnomad_sv2(
        InputFileType::GnomadSv2,
        "tests/db/to-bin/varfish-db-downloader/vardbs/grch37/strucvar/gnomad_sv.bed.gz"
    )]
    #[case::gnomad_cnv4(
        InputFileType::GnomadCnv4,
        "tests/db/to-bin/varfish-db-downloader/vardbs/grch38/strucvar/gnomad-cnv.bed.gz"
    )]
    #[case::gnomad_sv4(
        InputFileType::GnomadSv4,
        "tests/db/to-bin/varfish-db-downloader/vardbs/grch38/strucvar/gnomad-sv.bed.gz"
    )]
    #[case::inhouse_db(
        InputFileType::InhouseDb,
        "tests/db/to-bin/varfish-db-downloader/vardbs/grch37/strucvar/inhouse.tsv"
    )]
    fn test_deserialize_branch(
        #[case] input_type: InputFileType,
        #[case] path_input: &str,
    ) -> Result<(), anyhow::Error> {
        mehari::common::set_snapshot_suffix!(
            "{:?}-{}",
            input_type,
            path_input
                .split('/')
                .last()
                .unwrap()
                .split('.')
                .next()
                .unwrap()
        );

        let mut reader = csv::ReaderBuilder::new()
            .has_headers(false)
            .comment(Some(b'#'))
            .delimiter(b'\t')
            .from_reader(mehari::common::io::std::open_read_maybe_gz(path_input)?);

        let records = super::deserialize_branch(input_type, &mut reader)?;
        insta::assert_yaml_snapshot!(records);

        Ok(())
    }
}
