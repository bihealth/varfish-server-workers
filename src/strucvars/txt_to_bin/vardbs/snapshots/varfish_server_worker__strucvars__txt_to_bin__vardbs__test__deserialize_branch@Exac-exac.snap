---
source: src/strucvars/txt_to_bin/vardbs/mod.rs
expression: records
---
- svType: Dup
  start: 69089
  stop: 899565
  count: 1
- svType: Dup
  start: 69089
  stop: 949863
  count: 1
- svType: Dup
  start: 861263
  stop: 861409
  count: 1
- svType: Dup
  start: 861263
  stop: 861409
  count: 1
- svType: Dup
  start: 861263
  stop: 866474
  count: 1
- svType: Dup
  start: 861263
  stop: 871281
  count: 1
- svType: Dup
  start: 861263
  stop: 883988
  count: 1
- svType: Dup
  start: 861263
  stop: 892658
  count: 1
- svType: Dup
  start: 861263
  stop: 892658
  count: 1
- svType: Dup
  start: 861263
  stop: 900576
  count: 1
- svType: Dup
  start: 861263
  stop: 901999
  count: 1
- svType: Dup
  start: 861263
  stop: 901999
  count: 1
- svType: Dup
  start: 861263
  stop: 901999
  count: 1
- svType: Dup
  start: 861263
  stop: 909749
  count: 1
- start: 861263
  stop: 909960
  count: 1
- start: 861263
  stop: 909960
  count: 1
- svType: Dup
  start: 861263
  stop: 909960
  count: 1
- svType: Dup
  start: 861263
  stop: 909960
  count: 1
- svType: Dup
  start: 861263
  stop: 909960
  count: 1
- svType: Dup
  start: 861263
  stop: 909960
  count: 1
- svType: Dup
  start: 861263
  stop: 957847
  count: 1
- svType: Dup
  start: 861263
  stop: 970709
  count: 1
- start: 861263
  stop: 981650
  count: 1
- svType: Dup
  start: 861263
  stop: 985976
  count: 1
- svType: Dup
  start: 861263
  stop: 985976
  count: 1
- svType: Dup
  start: 861263
  stop: 985976
  count: 1
- svType: Dup
  start: 861263
  stop: 987200
  count: 1
- start: 861263
  stop: 1025811
  count: 1
- svType: Dup
  start: 861263
  stop: 1159353
  count: 1
- svType: Dup
  start: 871150
  stop: 909960
  count: 1
- svType: Dup
  start: 878631
  stop: 970709
  count: 1
- svType: Dup
  start: 879076
  stop: 907535
  count: 1
- svType: Dup
  start: 880896
  stop: 909960
  count: 1
- start: 883868
  stop: 889554
  count: 1
- start: 896671
  stop: 899565
  count: 1
- start: 898715
  stop: 900576
  count: 1
- start: 905655
  stop: 909960
  count: 1
- start: 905655
  stop: 909960
  count: 1
- start: 905655
  stop: 912007
  count: 1
- start: 905655
  stop: 912007
  count: 1
- svType: Dup
  start: 905655
  stop: 985976
  count: 1
- start: 908878
  stop: 909960
  count: 1
- svType: Dup
  start: 948952
  stop: 957847
  count: 1
- start: 976856
  stop: 985976
  count: 1
- svType: Dup
  start: 977334
  stop: 985976
  count: 1
- svType: Dup
  start: 1018272
  stop: 1025811
  count: 1
- svType: Dup
  start: 1102484
  stop: 1120527
  count: 1
- svType: Dup
  start: 1111010
  stop: 1111779
  count: 1
- svType: Dup
  start: 1117119
  stop: 1226079
  count: 1
- svType: Dup
  start: 1117119
  stop: 1237431
  count: 1
- svType: Dup
  start: 1117739
  stop: 1119476
  count: 1
- svType: Dup
  start: 1117739
  stop: 1120527
  count: 1
- svType: Dup
  start: 1117739
  stop: 1223221
  count: 1
- svType: Dup
  start: 1117739
  stop: 1256479
  count: 1
- svType: Dup
  start: 1117739
  stop: 1267323
  count: 1
- start: 1117739
  stop: 1451473
  count: 1
- svType: Dup
  start: 1139412
  stop: 1342602
  count: 1
- svType: Dup
  start: 1152890
  stop: 1164178
  count: 1
- svType: Dup
  start: 1152890
  stop: 1164178
  count: 1
- start: 1179569
  stop: 1444692
  count: 1
- svType: Dup
  start: 1190581
  stop: 1203377
  count: 1
- svType: Dup
  start: 1190581
  stop: 1203377
  count: 1
- start: 1198724
  stop: 1203377
  count: 1
- start: 1222146
  stop: 1236077
  count: 1
- svType: Dup
  start: 1222146
  stop: 1237431
  count: 1
- start: 1225649
  stop: 1226079
  count: 1
- start: 1225855
  stop: 1226525
  count: 1
- svType: Dup
  start: 1225855
  stop: 1451473
  count: 1
- svType: Dup
  start: 1233200
  stop: 1233498
  count: 1
- svType: Dup
  start: 1233391
  stop: 1249750
  count: 1
- svType: Dup
  start: 1237366
  stop: 1238664
  count: 1
- start: 1245967
  stop: 1260027
  count: 1
- svType: Dup
  start: 1245967
  stop: 1278143
  count: 1
- svType: Dup
  start: 1245967
  stop: 1278143
  count: 1
- svType: Dup
  start: 1245967
  stop: 1463247
  count: 1
- svType: Dup
  start: 1250202
  stop: 1251003
  count: 1
- start: 1262289
  stop: 1263148
  count: 1
- start: 1262289
  stop: 1289491
  count: 1
- start: 1262289
  stop: 1289491
  count: 1
- start: 1262289
  stop: 1342404
  count: 1
- svType: Dup
  start: 1267016
  stop: 1334056
  count: 1
- svType: Dup
  start: 1293834
  stop: 1461916
  count: 1
- svType: Dup
  start: 1309179
  stop: 1460676
  count: 1
- start: 1322609
  stop: 1342404
  count: 1
- start: 1323155
  stop: 1389883
  count: 1
- start: 1328160
  stop: 1337641
  count: 1
- svType: Dup
  start: 1328774
  stop: 1334691
  count: 1
- start: 1328774
  stop: 1337641
  count: 1
- svType: Dup
  start: 1333611
  stop: 1337641
  count: 1
- svType: Dup
  start: 1337461
  stop: 1342404
  count: 1
- svType: Dup
  start: 1337461
  stop: 1342404
  count: 1
- svType: Dup
  start: 1337461
  stop: 1342602
  count: 1
- svType: Dup
  start: 1337461
  stop: 1342602
  count: 1
- svType: Dup
  start: 1337461
  stop: 1342602
  count: 1
- svType: Dup
  start: 1342287
  stop: 1342404
  count: 1
- svType: Dup
  start: 1342287
  stop: 1342404
  count: 1
- svType: Dup
  start: 1342287
  stop: 1342404
  count: 1
- svType: Dup
  start: 1342287
  stop: 1342404
  count: 1
- svType: Dup
  start: 1342287
  stop: 1342404
  count: 1
